From 56d0acdf2112ccad55d04e2debe62cfcd48dbe96 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 5 Mar 2022 20:50:15 +0100
Subject: [PATCH 04/22] winex11.drv: Use set_focus in FocusIn events to
 activate focused window.

Setting the X11 input focus to the new focus window after calling
NtUserSetForegroundWindow, if necessary.

CW-Bug-Id: #19553
CW-Bug-Id: #19644
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=2155
---
 dlls/winex11.drv/event.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -594,7 +594,7 @@ static void set_input_focus( struct x11drv_win_data *data )
 static void set_focus( Display *display, HWND hwnd, Time time )
 {
     HWND focus, old_active;
-    Window win;
+    Window window;
     GUITHREADINFO threadinfo;
 
     old_active = NtUserGetForegroundWindow();
@@ -616,12 +616,12 @@ static void set_focus( Display *display, HWND hwnd, Time time )
     focus = threadinfo.hwndFocus;
     if (!focus) focus = threadinfo.hwndActive;
     if (focus) focus = NtUserGetAncestor( focus, GA_ROOT );
-    win = X11DRV_get_whole_window(focus);
+    if (focus == hwnd && !use_take_focus) return;
 
-    if (win)
+    if ((window = X11DRV_get_whole_window( focus )))
     {
-        TRACE( "setting focus to %p (%lx) time=%ld\n", focus, win, time );
-        XSetInputFocus( display, win, RevertToParent, time );
+        TRACE( "setting focus to %p/%lx, time %lu\n", focus, window, time );
+        XSetInputFocus( display, window, RevertToParent, time );
     }
 }
 
@@ -815,8 +815,9 @@ static BOOL X11DRV_FocusIn( HWND hwnd, XEvent *xev )
         if (!hwnd) hwnd = get_active_window();
         if (!hwnd) hwnd = x11drv_thread_data()->last_focus;
         if (hwnd && can_activate_window(hwnd)) set_focus( event->display, hwnd, CurrentTime );
+        return TRUE;
     }
-    else NtUserSetForegroundWindow( hwnd, FALSE );
+    else set_focus( event->display, hwnd, CurrentTime );
     return TRUE;
 }
 
-- 
2.36.0

